keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1811 @@
+integrating with them, although certain implementation details of the
+`iptables` backend in particular (e.g. the fact that service IPs in
+packets are rewritten to endpoint IPs during iptables's `PREROUTING`
+phase, and that masquerading will not happen before `POSTROUTING`) are
+effectively API, in that we know that changing them would result in
+significant ecosystem breakage.
+
+We should provide a stronger definition of these larger-scale "black
+box" guarantees in the `nftables` backend. NFTables makes this easier
+than iptables in some ways, because each application is expected to
+create their own table, and not interfere with anyone else's tables.
+If we document the `priority` values we use to connect to each
+nftables hook, then admins and third party developers should be able
+to reliably process packets before or after kube-proxy, without
+needing to modify kube-proxy's chains/rules.
+
+In cases where administrators want to insert rules into the middle of
+particular service or endpoint chains, we would have the same problem
+that the `iptables` backend has, which is that it would be difficult
+for us to avoid accidentally overwriting them when we update rules.
+Additionally, we want to preserve our ability to redesign the rules
+later to take better advantage of nftables features, which would be
+impossible to do if we were officially allowing users to modify the
+existing rules.
+
+One possibility would be to add "admin override" vmaps that are
+normally empty but which admins could add `jump`/`goto` rules to for
+specific services to augment/bypass the normal service processing. It
+probably makes sense to leave these out initially and see if people
+actually do need them, or if creating rules in another table is
+sufficient.
+
+```
+<<[UNRESOLVED external rule integration API ]>>
+
+It will be easier to figure out what the right thing to do here is
+once we actually have a working implementation.
+
+<<[/UNRESOLVED]>>
+```
+
+#### Rule monitoring
+
+Proxies based on iptables have historically had the problem that
+system processes (particularly firewall implementations) would
+sometimes flush all iptables rules and restart with a clean state,
+thus completely breaking kube-proxy. The initial solution for this
+problem was to just recreate all iptables rules every 30 seconds even
+if no services/endpoints had changed. Later this was changed to create
+a single "canary" chain, and check every 30 seconds that the canary
+had not been deleted, and only recreate everything from scratch if the
+canary disappears.
+
+NFTables provides a way to monitor for changes without doing polling;
+you can keep a netlink socket open to the kernel (or a pipe open to an
+`nft monitor` process) and receive notifications when particular kinds
+of nftables objects are created or destroyed.
+
+However, the "everyone uses their own table" design of nftables means
+that this may not be necessary. IPTables-based firewall
```

Comment:
There are a million things an administrator could do that would break kube-proxy; deleting the DaemonSet, removing the node IP, `down`ing the interface, adding bad routes, accidentally creating rules in other tables that act on k8s traffic, setting or unsetting various networking sysctls...

In the case of iptables, the issue was that if you used an iptables firewall _in the way that you were supposed to use it_, it would break kube-proxy. That was bad and we had to work around it. In the nftables case, there is no reason to think that this specific sort of breakage is going to happen, and thus no reason to pre-emptively try to protect against it.

Reply:
Agree that you can not stop disasters like `sudo rm -f` , but disagree that `nftables` has enough guardrails to prevent problems , is not just `iptables firewalls`, are implementations of network policies or service meshes, or network plugins with portmap, ...

Reply:
in order to keep our software reliable we should keep implementing the controller pattern with level based controllers https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#writing-controllers
```
for {
  desired := getDesiredState()
  current := getCurrentState()
  makeChanges(desired, current)
}
```
and not depend on `nft` to do the right thing, the important is the sum of all parts, if we can detect and warn or repair a situation we should do it, if the is a bug in the other component this will allow us to signal and detect it without disruption

Reply:
> is not just iptables firewalls

No, it _is_ just iptables firewalls. There is a very specific problem there which we had to work around, and we work around it very specifically. (`utiliptables.Monitor` does not protect against arbitrary iptables changes by users; it _only_ attempts to catch the case where every iptables rule gets deleted at the same time.)

> implementations of network policies or service meshes, or network plugins implementing portmap

If any of those things touch our nftables rules, then they are unambiguously broken. And I don't even just mean "they're using the nftables API in a naughty and unsupported way", I mean "they can't possibly be getting the behavior they want", because if their functionality depends on them changing our rules, it will break when change those rules back (as we will _eventually_, even if we don't notice the changes immediately).

(To be clear: unless we add some API a la the ["admin override" maps](https://github.com/kubernetes/enhancements/pull/3824/files#diff-85ab4fa1a983d0d1d33b6bc089dc2c5c0b558d73672e7e2afad8a647ad210061R1168) that are vaguely suggested in the KEP, then we would not support any user/component _ever_ making _any_ changes to the `kube-proxy` table. You do not need to make changes or additions to the `kube-proxy` table to implement NetworkPolicy or portmaps; you make _your own_ tables and put your rules there.)

> in order to keep our software reliable we should keep implementing the controller pattern with level based controllers

Absolutely, we should not assume that kube-proxy manages to observe all events, or that it is always running.

But we intentionally don't stick _strictly_ to the level-based controller pattern in iptables kube-proxy, because, given the APIs available to us and the amount of data we have to process, that would result in unacceptable levels of performance. Instead, we do a slightly-looser sync loop, with periodic "overwrite everything with the expected state" fallbacks. The nftables backend will presumably work the same way.

(Or maybe the nftables APIs will turn out to be efficient enough that we can do a stricter sync instead... the big problem with iptables is that `getCurrentState()` is almost as slow as `makeChanges(desired, nil)` would be.)

> if we can detect and warn or repair a situation we should do it

There is a maintenance burden to having code to detect/warn/repair, so we shouldn't do it unless we think it is _actually_ going to be a problem. At the current time, we have no reason to believe that people deleting our nftables rules out from under us is going to be a problem.

Reply:
> There is a maintenance burden to having code to detect/warn/repair, so we shouldn't do it unless we think it is actually going to be a problem. At the current time, we have no reason to believe that people deleting our nftables rules out from under us is going to be a problem.

I see, ok, agree with most of your points, I'm not advocating for doing the reconcile each time, but I really believe we should  have a repair loop that run every X mins that do the diff, like we do with the allocators, this loop has proven to be fantastic all these years to improve the resilience of the apiserver, if we add metrics to this loop like in https://github.com/kubernetes/kubernetes/pull/120843,  this will allow us to confirm with data the hypothesis that it is not really needed ...

Reply:
Any repair / reconcile for **alpha** should aim to be as simple as we can make it (and any spare capacity dedicated to help with migration, docs, supporting tooling).

For beta and certainly before GA we should aim to make the reconcile mechanism work well even for high throughput and large scale - so no “fetch and compare all nftables records every inner invocation” shortcuts.

As this KEP PR aims to be `implementable` at alpha, I think we only need a nod / `UNRESOLVED` paragraph about what we might do for later phases. Let's [DoTheSimplestThingThatCouldPossiblyWork](https://wiki.c2.com/?DoTheSimplestThingThatCouldPossiblyWork) for our first pass at alpha.

Reply:
> I really believe we should have a repair loop that run every X mins that do the diff

We handle "repair"; we just don't do it with a separate loop or a diff. `syncProxyRules` just completely overwrites the current state with the desired state sometimes, to ensure that if any errors have crept in, they get fixed.

In the iptables code, there is a separate goroutine that polls the canary chains and forces an immediate "repair" (aka "full sync") if they disappear. (It has to do this in a slightly-weird way, to minimize the amount of time it holds the xtables lock, to avoid slowing down the main sync thread in the normal case where the canaries _haven't_ been deleted.) In the nftables code, I don't plan to have that check, since we don't expect that failure mode, but we'll still have the _periodic_ "completely overwrite the current state with the desired state", as well as the "completely overwrite the current state with the desired state" if we ever get back an error from nftables (eg it claims a chain doesn't exist when we think it should).

Reply:
OK, I think I see what you were asking in the other thread about reading back rules and diffing.

In the iptables backend, the only diff-like thing we ever do is that we read back the set of chain names that currently exist, so we can compare them against the set of chains we think _should_ exist, and delete any stale ones that were left behind.

Other than that, we _never_ compare the existing rules to the desired rules. If we want to make any change to an iptables chain, we just tell iptables "delete all of the existing rules in this chain and set the rules to be these new ones". (And if we _don't_ want to make a change to a chain, then we just ignore it, until either we get an error, or a timeout occurs, at which point we decide that we want to change _every_ chain all at once.)

(The only reason we don't do the same thing for deleting stale chains is that iptables doesn't give us any API that would let us do that, unless we wanted to delete all non-Kubernetes chains too.)

The nftables backend will likely work exactly the same, in this regard, except that there may be more things we need to delete besides just chains (eg, sets and maps).